### DISCUSSION TITLE
How to add custom Icons and and custom UI elements ?

### DISCUSSION THREAD

**Discussion by faronov** [zigpy/zha-device-handlers#2992] [Q&A] [answered]
Hi, I'm trying to add support for some custom attributes to the UI configuration of a device. 
I'm able to create quirck and add desired attributes to the cluster. 
I can also able set/get attribute in the "Manage Zigbee devices" view. 
However, I can't find any documentation (or I clearly don't understand) on how to display these attributes in the UI and add some configuration elements like switches, sliders, or input fields, along with specific text.
I've read some discussions and ideas about "exposes," but this is still an ongoing discussion. 
Can you suggest a starting point to understand how to add icons and fields/switches/sliders to the UI?

**Comment by Hedda** [comment]
To expose new device types or new clusters and attributes that are not already exposed in the UI of ZHA you need modify the zha component code inside Home Assistant’s core -> https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha

**Comment by Hedda** [comment]
Depending on what you want to expose it might also be good to also post such questions to the zigpy reposoitory as more ZHA developers are likley to spot such questions if create discussion there -> https://github.com/zigpy/zigpy

> I've read some discussions and ideas about "exposes," but this is still an ongoing discussion.

Btw, I believe that discission you are referring to can be found here -> https://github.com/zigpy/zigpy/discussions/1312

**Reply by faronov** [reply]
Yes I'm referring to that discussion 

This just some additional settings . like turn on/off led or configure led blink time , this is completely not functional , and not compliant at all with ZigBee standard ,  and not so important that must be extended in zha component
I just think , that maybe some quirk  trick exist to add such custom settings to UI

**Comment by TheJulianJES** [comment, marked as answer]
ZHA needs to be modified for now, but we're adding support for "[quirks v2](https://github.com/zigpy/zigpy/discussions/1312#discussioncomment-8191961)" soon (if everything goes as planned).
You'll be able to expose attributes for ZHA to automatically create switches/sensors/...

**Comment by Hedda** [comment]
> ZHA needs to be modified for now, but we're adding support for "[quirks v2](https://github.com/zigpy/zigpy/discussions/1312#discussioncomment-8191961)" soon (if everything goes as planned).
> You'll be able to expose attributes for ZHA to automatically create switches/sensors/...

FYI, ZHA/zigpy developers have now introduced initial support for what they call ”**_quirks v2_**” and dmulcahey begun how-to documentation:

https://github.com/zigpy/zha-device-handlers/pull/3019

This is a huge change since will allow adding new entities to Home Assistant via quirks alone without having to modify ZHA code in HA core.

For more information see these for reference:

https://github.com/zigpy/zigpy/discussions/1312

https://github.com/zigpy/zigpy/pull/1329

https://github.com/home-assistant/core/pull/111176

That PR with new docs will initially add a separate `quirks_v2.md` file to difference v1 from quirks v2 -> https://github.com/zigpy/zha-device-handlers/pull/3019

```
# Quirks v2

## Introduction

Quirks v2 use a fluent interface style. While this isn't common in python it was done to improve the readability of the source code and to make it human-friendly for non developers. The amount of boilerplate code has been significantly reduced and the need to specify a signature dictionary and a replacement dictionary has been removed. This should make it much easier for the community to contribute quirks.

## Goals

- Significantly reduce the amount of boilerplate code required to write a quirk
- Make it easier for the community to contribute quirks
- Make it easier to read and understand quirks
- Make it easier to maintain quirks
- Expose entities from a quirk
- Allow custom logic to determine if a quirk should be applied to a device
- Allow custom binding, reporting configuration or any sort of initialization logic without hacking the bind or configure_reporting methods

## Breakdown of a minimal example

...

```

PS: Also check out summarized back-story and more links here:

https://community.home-assistant.io/t/zha-device-handlers-version-2-zha-quirks-v2-architectural-design-developer-discussion/666460
